### PR TITLE
[DEV-2030] UMA

### PR DESCRIPTION
**Pull request by bernd2022**
This version is not yet finalized. It is the current implementation status, which works to a large extent.

### CODE REVIEW

File: src/integration/blockchain/uma/services/uma.service.ts [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+  async createUmaInvoice(amountMsats: number, metadata: string): Promise<string | undefined> {
+    const receiverMetadata = metadata.slice(0, metadata.indexOf('}{') + 1);
+    const receiverMetadataAsJson = JSON.parse(receiverMetadata);
+    const lnReceiverAddress = (<string>receiverMetadataAsJson.receiver).replace('$', '');
+
+    const lnPayRequest = await this.lightningForwardService.wellknownForward(lnReceiverAddress);
+    const lnCallback = lnPayRequest.callback;
+    const lnUrlpId = lnCallback.slice(lnCallback.lastIndexOf('/') + 1);
+
+    return this.lightningForwardService.lnurlpCallbackForward(lnUrlpId, { amount: amountMsats }).then((i) => i.pr);
```
Comment (davidleomay) [inline src/integration/blockchain/uma/services/uma.service.ts:322]: Could this be done in a simpler way?

File: src/integration/blockchain/uma/services/uma.service.ts
Comment (davidleomay) [inline src/integration/blockchain/uma/services/uma.service.ts:1]: Can we structure this file into `Sender Methods`, `Receiver Methods` and shared `Helper Methods`?

File: src/integration/blockchain/uma/services/uma.service.ts [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+  constructor(
+    readonly http: HttpService,
+    private readonly lightningForwardService: LightningForwardService,
+    private readonly coingeckoService: CoinGeckoService,
+  ) {
+    this.umaClient = new UmaClient(http);
+
+    // InMemoryNonceValidator should not be used in production!
+    this.nonceValidator = new InMemoryNonceValidator(1000);
```
[the remark this answers is not in the document]
Comment (bernd2022) [reply src/integration/blockchain/uma/services/uma.service.ts:47]: Currently during test we can use it. For the final Version we could implement a nonce validation with help of a database table storage

File: src/integration/blockchain/uma/services/uma.service.ts [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    return LightningHelper.btcToMsat(base / price);
+  }
+
+  private async getPriceInBTC(from: string): Promise<number> {
+    //const price = await this.coingeckoService.getPrice(from, 'btc');
+    //if (!price.isValid) throw new InternalServerErrorException(`Invalid price from ${from} to btc`);
+
+    //return price.price;
+    return from === 'usd' ? 66500 : 62500;
+  }
```
[the remark this answers is not in the document]
Comment (bernd2022) [reply src/integration/blockchain/uma/services/uma.service.ts]: Maybe the price cache could be valid as long, as the invoice is valid - just a spontaneous idea

File: infrastructure/bicep/lightning-api.bicep
Comment (davidleomay) [inline infrastructure/bicep/lightning-api.bicep:1]: CoinGecko API key is missing

**Pull request merged**